### ISSUE TITLE
Moving tabs between groups very slow

### ISSUE THREAD

**Issue by diego898** [josh-berry/tab-stash#305]
### Question

I recently imported ~1000 tabs into tab-stash into one group. This understandably took a long while. Now, I am splitting this large group into others - for example I have a "Hacker News" group. There are about ~400 tabs in my import group that I want moved. Searching is fast and I select all of them and click to move them to the other group. However, I can see each URL processing/moving and it takes about half a second. This seems very long. 

Questions
1. Isn't moving urls between groups just a text operation? 
2. How can I ask tab stash to "fetch missing title" and "Favicons" for my imported urls? 

### System Details

- **OS Version:** Mac OSX
- **Browser Version:** 107.0
- **Tab Stash Version:** 2.11.1

**Comment by josh-berry** [comment]
>     1. Isn't moving urls between groups just a text operation?

Nope, it's a database operation. Unfortunately I can't tell you why it takes so long, because when I've measured it in the past, almost all the time is spent inside Firefox itself. Bookmark operations in general seem to be rather slow on Firefox, for reasons that are beyond me.

>     2. How can I ask tab stash to "fetch missing title" and "Favicons" for my imported urls?

For favicons, there is a menu item "Fetch Missing Icons" in the Tab Stash menu next to the search bar.

However, there's not a way to reset/fetch titles in bulk right now unfortunately. It's possible to do one by one if you load the tab, edit the title in the stash, and clear the text box (so it gets the default title again, which is just the page title).

The other option would be to export and re-import the tabs that don't have correct titles—the importer should pick the correct icon and title.

Your question makes me curious, though—was the importer NOT able to fetch the correct icon/title for some of your imported tabs? And if not, can you share any examples so I can take a closer look?

**Comment by josh-berry** [comment]
> Ah strange - I never considered this was using Firefox bookmarks (always thought it was just a list of urls in a text file, or similar). Could firefox syncing be an issue then?

Syncing happens asynchronously so shouldn't be a problem here. AFAICT it's really the database update itself that's the problem. Firefox uses a sqlite database to store its bookmarks, history, etc. and it seems to force a full transaction commit, flush to disk, etc. on every single extension API call.

(Syncing does cause other fun problems, though! Some of the worst Tab Stash bugs—long since fixed, of course—have actually been caused by Firefox Sync happening at precisely the wrong time...)

> Is it faster to just write a large batch of bookmarks instead of one by one?

It almost certainly would be, if Firefox gave us APIs for that. 😆 Unfortunately all the bookmark APIs only let you touch one bookmark at a time.

And it's true the bookmark APIs are async APIs... but doing lots of bookmark operations in parallel (especially move operations) triggers some fun bugs in Firefox itself. For example, I've seen bugs where bookmarks get mis-ordered in a way that prevents you from re-ordering them (as in, you try to drag them back to their proper locations and they just don't go where they're supposed to, or other bookmarks get shuffled, or ... you get the idea).

Long story short, performance of Firefox's bookmark API isn't great, but every time I've tried to work around it in some clever way, it's usually caused bugs either in Tab Stash or in Firefox itself.

> At this point, I'm considering exporting all groups, manipulating all urls in a text editor to move them between groups, and then just re-importing all of them.

It might also be faster to reorganize your bookmarks using the built-in bookmark editor itself. For whatever reason (probably because it has access to better internal APIs), moving batches of bookmarks around is faster there.

One of the benefits of using a standard storage medium like bookmarks is you can use tools besides Tab Stash to keep your stash organized. ;)

> As for 2. I was importing about ~500 links from news.ycombinator.com (HackerNews). The first time I did this I think multi-account containers made the import fail all over the place and I only realized this after it finished so I deleted everything and did it again with all extensions except tab stash disabled.

hrm... there's code in the importer explicitly to deal with multi-account containers. If a tab gets replaced, the importer should notice and pick up the new tab and everything should just work.

If you're inclined to debug this further and want to open a bug report (no worries if not), here's how to get some debug info:

1. Open Tab Stash in its own tab (right-click -> ... -> "Show Stashed Tabs in a Tab")
2. Open the JavaScript console and run `trace_siteinfo = true`
3. Import a few URLs which are pinned to particular containers (or whichever URLs were causing you trouble before)
4. Get lots of debug output

> Also - is it possible to request a batch fetch updated titles in this group button? Perhaps with a built-in cache that persists across sessions so that it can happen slowly over a period of time to prevent rate limiting/locking up the browser?

Sure! Given that we've gone back and forth in this issue on a couple of your questions already, would you mind opening a separate issue for it? That way I can keep things organized and we can dig into that feature request in more detail in its own thread.

> Anyway, just thinking out loud, and I don't mean to come off like I'm just demanding more and more! I very much appreciate this great plugin and all your efforts!

No worries, happy to help!

**Comment by diego898** [comment]
Great! I made a new one, and will close this issue
